cram/cram_codecs.c

```diff
@@ -1050,24 +1047,32 @@ cram_codec *cram_huffman_decode_init(char *data, int size,
 	if (max_len < codes[i].len)
 	    max_len = codes[i].len;
     }
-    if (l < 1 || cp - data != size || max_len >= ncodes) {
-	hts_log_error("Malformed huffman header stream");
-	free(h);
-	return NULL;
+    if (l < 1 || cp - data != size || max_len >= ncodes)
+        goto malformed;
+
+    /* 31 is max. bits available in val */
+    if (max_len > max_code_bits) {
+        hts_log_error("Huffman code length (%d) is greater "
+                      "than maximum supported (%d)", max_len, max_code_bits);
+        free(h);
+        free(codes);
+        return NULL;
     }
 
     /* Sort by bit length and then by symbol value */
     qsort(codes, ncodes, sizeof(*codes), code_sort);
 
     /* Assign canonical codes */
-    val = -1, last_len = 0;
+    val = -1, last_len = 0, max_val = 1;
```

Comment:
About the only place huffman tables are now generated in htslib is for zero length codes as we have no CRAM codec capable of stating something is a constant.  A huffman tree of one symbol (zero length) is the closest we get.

What if we have two symbols claiming zero length?  At this point we have two loops through here with val 0, val 1, but the second one doesn't trigger "val > max_val" error.  Should max_val be 0 to start with?

PS.  The rest of the PR looks fine.